src/core/server/opensearch_data/client/data_source_client.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+import { Client } from '@opensearch-project/opensearch';
+import { Logger } from '../../logging';
+import { OpenSearchClient, OpenSearchClientConfig } from '../../opensearch/client';
+import { SavedObjectsClientContract } from '../../saved_objects/types';
+import { CryptoCli } from '../../../../../src/plugins/credential_management/server/crypto/cli/crypto_cli';
+
+/**
+ * TODO: update doc
+ * Represents an OpenSearch cluster API client created by the platform.
+ * It allows to call API on behalf of the internal OpenSearch Dashboards user and
+ * the actual user that is derived from the request headers (via `asScoped(...)`).
+ *
+ * @public
+ **/
+export interface IDdataSourceClient {
+  /**
+   * TODO: update doc. Creates a {@link IScopedClusterClient | scoped cluster client} bound to given {@link ScopeableRequest | request}
+   */
+  asDataSource: (dataSourceId: string) => Promise<OpenSearchClient>;
+}
+
+/**
+ * See {@link IClusterClient}
+ *
+ * @public
+ */
+export interface ICustomDataSourceClient extends IDdataSourceClient {
+  /**
+   * Closes the data source client. After that client cannot be used and one should
+   * create a new client instance to be able to interact with OpenSearch API.
+   */
+  close: () => Promise<void>;
+}
+
+// TODO: tmp
+interface DataSourceInfo {
+  endpoint: string;
+  credentialId: string;
+}
+
+interface CredentialInfo {
+  username: string;
+  password: string;
+}
+
+export class DataSourceClient implements ICustomDataSourceClient {
+  public dataSourceClientsPool: Map<string, Client>;
```

Comment:
need to limit the size of the Map

Reply:
You are right. Not only will I need to add size, but also build logic around it to pop-up least recent used dataSource client. For now I'll update this PR with a fixed size of 50. And added todos